### PR TITLE
add ckeditor media embed plugin

### PR DESCRIPTION
**Pull request by aliceriot**
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
  - [ ] Mobile width screenshots
  - [ ] Tag @ferdi or @pdpinch for review
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested
- [x] Settings
  - [x] New settings are documented and present in `app.json`
  - [x] New settings have reasonable development defaults, if applicable
  - [x] Opened issue for DevOps regarding necessary configuration changes to deployed environments

#### What are the relevant tickets?

for #39 

#### What's this PR do?

This adds the [MediaEmbed](https://ckeditor.com/docs/ckeditor5/latest/features/media-embed.html) plugin to our CKEditor setup. In order to support this we also need to extend our Markdown layer, so I added a new framework for doing that involving a new type of plugin.

It is configured right now to support only YouTube embeds - if we want to support more providers we'll need to extend the Markdown translation part to support other shortcodes.

#### How should this be manually tested?

You should be able to add a YouTube video to the editor. It should show up in the output as a Hugo-compatible shortcode. If you start the editor with a youtube shortcode in the input data it should up in the editor as a preview without any issues.

Also! Definitely review the architecture carefully around customizing the markdown layer. It's set up so each custom plugin will now also define a plugin that extends `MarkdownSyntaxPlugin`, and this plugin basically just defines the showdown and turndown rules that we want to use for the plugin. This will let us keep the code for the markdown stuff right with the rest of the code for the plugin, and will also ensure that markdown support is added to the editor at the same time as the rest of the new plugin.

#### Screenshots (if appropriate)

Screenshot from 2021-03-04 15-30-19
Screenshot from 2021-03-04 15-30-25

### CODE REVIEW

**Review by gumaerc — CHANGES_REQUESTED** [review]
The structure of the MarkdownSyntaxPlugin looks straightforward enough and matches what you talked about in your demo.  I raised one point about the regex matching that I think needs to be addressed, but other than that I think this looks good.  I didn't know about some of these turndown quirks that you are handling here, maybe we should add these rules to `ocw-to-hugo` as well?

File: static/js/lib/ckeditor/plugins/MarkdownMediaEmbed.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import Showdown from "showdown"
+
+import MarkdownSyntaxPlugin from "./MarkdownSyntaxPlugin"
+import { TurndownRule } from "../../../types/ckeditor_markdown"
+
+export const YOUTUBE_SHORTCODE_REGEX = /{{< youtube (\S+) >}}/
+
+export const EMBED_CLASS = "media"
+
+export const YOUTUBE_URL_REGEX = /https:\/\/www\.youtube\.com\/watch\?v=([a-zA-Z0-9_]+)\/?$/
```
[the remark this answers is not in the document]
Comment (gumaerc) [reply static/js/lib/ckeditor/plugins/MarkdownMediaEmbed.ts:13]: Also, I think any YouTube link that someone copies directly from YouTube should use `https`, but maybe we should make the s optional to be safe.

[the remark this answers is not in the document]
Comment (gumaerc) [reply static/js/lib/ckeditor/plugins/MarkdownMediaEmbed.ts:13]: Also, I think `www.` should be optional.

[the remark this answers is not in the document]
Comment (aliceriot) [reply static/js/lib/ckeditor/plugins/MarkdownMediaEmbed.ts:13]: cool, I'll polish up the regex and add some tests

[the remark this answers is not in the document]
Comment (aliceriot) [reply static/js/lib/ckeditor/plugins/MarkdownMediaEmbed.ts:13]: I updated it to account for those complications and also to support the short links, lmk what you think

File: static/js/lib/ckeditor/plugins/MarkdownMediaEmbed.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+          node: Turndown.Node,
+          _: Turndown.Options
+        ): string => {
+          const videoId = node
+            ?.querySelector("oembed")
+            ?.getAttribute("url")
+            ?.match(YOUTUBE_URL_REGEX)
+
+          if (videoId && videoId[1]) {
+            return `{{< youtube ${videoId[1]} >}}`
```
Comment (gumaerc) [inline static/js/lib/ckeditor/plugins/MarkdownMediaEmbed.ts:48]: This is a bit of a nitpick, but I noticed if you put two videos side by side without a line break in between, it renders the shortcodes on the same line:

```
{{< youtube fB9GlZUYNUQ >}}{{< youtube O_XhRdWdphU >}}
```

This seems to render fine in Hugo, I tested it by injecting this into a running course:

image

Alternatively, if you insert a line break after the first video, three line breaks seem to be inserted between the shortcodes:

```
{{< youtube fB9GlZUYNUQ >}}

 

{{< youtube O_XhRdWdphU >}}
```

This markdown renders exactly the same in Hugo, so there's no big reason to fix this aside from potential readability.  Ideally there should only be a single line break between the two videos.

**Comment by gumaerc** [comment]
@aliceriot Just one more question, feel free to punt on it for now as it's not super important.

File: static/js/lib/ckeditor/plugins/MarkdownMediaEmbed.ts (hunk 2 above)
Comment (aliceriot) [reply static/js/lib/ckeditor/plugins/MarkdownMediaEmbed.ts:48]: I added a `\n` at the end of the turndown rule, ensure that youtube videos right next to each other in the editor are on different lines:

Screenshot from 2021-03-09 14-53-08

**Review by gumaerc — APPROVED** [review]
:+1: LGTM

**Pull request merged**